src/main/java/com/emt/service/MatchService.java

```diff
@@ -65,4 +71,23 @@ public BigDecimal calculateProbability(BigDecimal rating1, BigDecimal rating2) {
 
     return ONE.divide(divisor, 2, HALF_UP);
   }
+
+  public void cancelMatch(Long matchId) {
+    Match matchToCancel =
+        matchRepository.findById(matchId).orElseThrow(() -> new MatchNotFoundException(matchId));
+
+    recalculateEloRatings(
+        matchToCancel.getWinner(), matchToCancel.getLoser(), matchToCancel.getWinnerRatingChange());
+
+    matchRepository.delete(matchToCancel);
+  }
+
+  private void recalculateEloRatings(Player winner, Player loser, BigDecimal winnerRatingChange) {
+    if (winnerRatingChange != null) {
+      winner.setEloRating(winner.getEloRating().subtract(winnerRatingChange));
+      loser.setEloRating(loser.getEloRating().add(winnerRatingChange));
+    }
+
+    playerRepository.saveAll(List.of(winner, loser));
+  }
```

Comment:
#16  Recalculation means not for the selected match only, but for all the subsequent as well